### DISCUSSION TITLE
dxtbx: issue with pycbf

### DISCUSSION THREAD

**Discussion by picca** [dials/dials#2198] [General]
Hello, I am sill trying to obtain a working dials stack on Debian :).

first I do not know which dxtbx repository (the one from cctbx of your) is the official repository of dxtbx. I can not fill issue about dxtbx on your repository, but the cctbx is a fork of your. So I am wondering which one should be use in order to work. today it seems that both are in sync.

Now I am coming back with the string / bytes issue due to pycbf.

I packaged the latest 0.9.7 version of cbflib and use the pycbf module (which is the Python3 version), and indeed I got this kind of error

```python

    def test_beam():
        image = str(Path(__file__).parent / "phi_scan_001.cbf")
>       assert BeamFactory.imgCIF(image)

tests/test_beam.py:10: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dxtbx/model/beam.py:259: in imgCIF
    cbf_handle.read_widefile(cif_file.encode(), pycbf.MSG_DIGEST)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pycbf.cbf_handle_struct; proxy of <Swig Object of type 'handle *' at 0x7f08b6b764b0> >
filename = b'/home/experiences/instrumentation/picca/debian/science-team/dxtbx/.pybuild/cpython3_3.9_dxtbx/build/tests/phi_scan_001.cbf', headers = 8

    def read_widefile(self, filename, headers):

[...]

>       return _pycbf.cbf_handle_struct_read_widefile(self, filename, headers)
E       TypeError: in method 'cbf_handle_struct_read_widefile', argument 2 of type 'char *'

/usr/lib/python3/dist-packages/pycbf.py:2086: TypeError
```

the encoding is not required anymore. it seems that pycbf expect now real python3 string and not byte.

I am a little bit confuse, and I would like to know how you make all this work together.

thanks for considering.

Frederic

**Comment by ndevenish** [comment]
> today it seems that both are in sync.

Yes, they are the same. For historical reasons, all the issues and PRs go against https://github.com/cctbx/dxtbx

> I packaged the latest 0.9.7 version of cbflib

Ah, I don't think any of us were aware that a 0.9.7 had been released. We were told that CBFlib development was moving to https://github.com/cbflib/cbflib which doesn't have a release, or any new commits left. DIALS has certainly never been built against 0.9.7 from here, it still uses 0.9.6. In fact, the 0.9.7 development branch PR appears to still be open and have more recent commits than the release date https://github.com/cbflib/cbflib/pull/1 ?

DIALS as-is used to have CBFlib to be built with `SWIG_PYTHON_STRICT_BYTE_CHAR`. In fact, to get around this and to release the conda-forge build of DIALS, we split off the pycbf parts into https://github.com/dials/pycbf, and added the python3 compatibility changes - this meant that pycbf could accept transparently bytes and strings. We were told that it was of prime importance that every existing use case would build and work as-is with CBFlib 0.9.7, so I am surprised that it doesn't work automatically.

In any case, probably there is work to be done to reconcile 0.9.7 CBFlib into DIALS.

**Comment by picca** [comment]
> &gt; I packaged the latest 0.9.7 version of cbflib

I am using this repository https://github.com/yayahjb/cbflib which contain a 0.9.7 tag.

It would be great effectively that I know wich repository is the right one. Thsi way I should avoid this kind of issue.

> Ah, I don&#039;t think any of us were aware that a 0.9.7 had been released. We were told that CBFlib development was moving to https://github.com/cbflib/cbflib which doesn&#039;t have a release, or any new commits left. DIALS has certainly never been built against 0.9.7 from here, it still uses 0.9.6. In fact, the 0.9.7 development branch PR appears to still be open and have more recent commits than the release date https://github.com/cbflib/cbflib/pull/1 ?

> DIALS as-is used to have CBFlib to be built with `SWIG_PYTHON_STRICT_BYTE_CHAR`. In fact, to get around this and to release the conda-forge build of DIALS, we split off the pycbf parts into https://github.com/dials/pycbf, and added the python3 compatibility changes - this meant that pycbf could accept transparently bytes and strings. We were told that it was of prime importance that every existing use case would build and work as-is with CBFlib 0.9.7, so I am surprised that it doesn&#039;t work automatically.

MAybe I have the wronf cbflib :(

> In any case, probably there is work to be done to reconcile 0.9.7 CBFlib into DIALS.

It would be great to have only one cbflib to care about.

Cheers

**Comment by ndevenish** [comment]
No, you have the right CBFlib. It's just DIALS is currently only built against the older 0.9.6 version, with cctbx' cbflib_adaptbx.

We did make changes in DIALS so that it could be added as a conda-forge package, but to work without cbflib_adaptbx (which wasn't present in the cctbx package), we added changes to let pycbf work without them. That, and tidying up the build so that pycbf could be stand-alone installed, was why https://github.com/dials/pycbf came about.

I can't find a way to build cctbx with the latest cbflib, all of the cctbx bootstrap methods I knew appear to be broken (either dead URLS or tqdm errors from miniforge), so I'm building a fresh DIALS to see if the new CBFlib works with it now.

In fact, looking at the code, it looks like instead of resolving the byte/str issue the same way we did, it looks like he is manually injecting `SWIG_PYTHON_STRICT_BYTE_CHAR` into the generated build to work around the issue?
https://github.com/yayahjb/cbflib/blob/5232590ec6cc05bb2705acc3413ea598263854da/pycbf/make_pycbf.py#L633

**Comment by ndevenish** [comment]
In fact, for a fresh dials bootstrap, manually updating to CBFlib 0.9.7 works perfectly:
```
git clone git@github.com:dials/dials modules/dials
git clone https://github.com/yayahjb/cbflib -b CBFlib-0.9.7 modules/cbflib
python3 modules/dials/installer/bootstrap.py base update build
source ./dials
pytest -n auto modules/dxtbx --regression
```
So, whatever the cctbx parts of the cbflib build are running appear to be compiling it the same way, it appears to be your build that is different.

Perhaps it is requiring `SWIG_PYTHON_STRICT_BYTE_CHAR`. On our intermediate pycbf repository I added a commit that let anything accepting str or bytes to accept both: https://github.com/dials/pycbf/pull/4/files, or there was my PR https://github.com/yayahjb/cbflib/pull/19 to resolve this directly, but it looks like the approach I noted above was used instead. Historically, there has been opposition to switching to non-strict pycbf, but that might be resolved now that it's been forced on for.... parts of the build. Still, migrating dxtbx to non-strict is probably some medium body of work.

**Comment by ndevenish** [comment]
In any case, since this seems to be a problem in your custom build, I'm converting this to a discussion rather than an issue.

**Comment by picca** [comment]
Hello, thanks for your advices. Since I am preparing a Debian package for dials. I started from the cctbx package

 you can find the packaging here:

https://salsa.debian.org/science-team/cctbx

this is not the latest but I hope  recent enought. I did not activated the cbflib_adaptbx in this build.

then I worked on the dxtbx package which is here

https://salsa.debian.org/science-team/dxtbx

the only patch applyed was

https://salsa.debian.org/science-team/dxtbx/-/blob/master/debian/patches/0001-added-missing-hdf5-dependency.patch

the build process went fine, but when I tryed to run the test with pytest and it end up with the error reported before.

I did no modification to pycbf from the upstream, so I am confuse now when you says that it works out of the box.

**Comment by ndevenish** [comment]
> I did not activated the cbflib_adaptbx in this build.

DIALS needs both cbflib_adaptbx and cbflib, if using the original sources. Since the conda-forge package for cctbx does not include cbflib_adaptbx, we worked around it - but only when using our modified pycbf. If you are building the packaging yourself, then you need to either:
1. Build cbflib_adaptbx and cbflib the same way cctbx does
2. Patch the dxtbx sources to compensate
3. Use an enhanced version of pycbf that avoids the problem.
4. Use some other method for the missing parts of cbflib e.g. a replacement for what we are doing in [FormatMarIP](https://github.com/cctbx/dxtbx/blob/e4bc7c4965cfbb7e43df46e9ddea56a2bbb27119/src/dxtbx/format/FormatMarIP.py#L13-L27).

ISTR that CBFlib itself did have some of the functionality missing from pycbf, which we also bind to in [cbf_read_buffer.h](https://github.com/cctbx/dxtbx/blob/main/src/dxtbx/format/boost_python/cbf_read_buffer.cpp) (or use if pycbf has it e.g. [here](https://github.com/cctbx/dxtbx/blob/e4bc7c4965cfbb7e43df46e9ddea56a2bbb27119/src/dxtbx/format/FormatCBFFull.py#L22-L29)). This might have been fixed for CBFlib 0.9.7.

> I did no modification to pycbf from the upstream, so I am confuse now when you says that it works out of the box.

The canonical way to build DIALS is with the `bootstrap.py`, which builds cctbx, cbflib and cbflib_adaptbx. The above script is what I ran to build DIALS and dxtbx with CBFlib 0.9.7 - this appeared to work fine. If building everything from scratch, that is the "box". Outside of that, if you are building things in a different way then of course you are likely to run into problems that we don't normally see.

**Comment by picca** [comment]
Thanks a lot for these informations. In the Debian world, I need to package each project one by one. For security reasons, the network is not available during the build process. So I need to have all the sources available in order to process the build. The difficulty I am facing now is that I need the cbflib source files in order to build the cbflib_adaptbx. I think that I will need to patch the cctbx build system in order to avoid this probleme and directly link to the libcbf1 sources. Do you know is cctbx are modifying the cbflib during this process ?

**Reply by ndevenish** [reply]
I know that some of the cbf packages have historically done source rewriting on dependent packages, but don't think cbflib_adaptbx is one of them. I believe their build is controlled here: https://github.com/cctbx/cctbx_project/blob/master/cbflib_adaptbx/SConscript.

**Reply by picca** [reply]
Thanks, I will try to build cbflib_adaptbx from cctbx

Cheers

**Reply by picca** [reply]
In fact the SConscript of cbflib_adaptbx rebuild the `pycbf_wrap.c` and modify it by passing the  `-DSWIG_PYTHON_STRICT_BYTE_CHAR` flag. So this is not compatible with the current pycbf  module available in Debian. Nevertheless I will try to build the cbflib_adaptbx link with the libcbf and see if it works.

```python
 env_cbflib_boost_python_ext = env_no_includes_boost_python_ext.Clone()
  env_cbflib_boost_python_ext.Prepend(CPPFLAGS=["-DSWIG_PYTHON_STRICT_BYTE_CHAR"])

  env_etc.include_registry.append(
    env=env_cbflib_boost_python_ext,
    paths=env_etc.cbflib_common_includes + [env_etc.python_include])

  env_cbflib_boost_python_ext.SharedLibrary(
    target='#lib/_pycbf',
    source=[prefix+"/pycbf/pycbf_wrap.c",
            prefix+"/src/cbf_airy_disk.c"],
    LIBS=["cbf"]+env_etc.libs_python+env_etc.libm)

  Export("env_cbflib_boost_python_ext")
 ```

**Reply by ndevenish** [reply]
Yes. Before 0.9.7, python 3 wasn't really supported by CBFlib. You _had_ to set `SWIG_PYTHON_STRICT_BYTE_CHAR`, because your options were a) everything bytes, including strings - mildly annoying, or b) everything strings, including large data arrays as strings - and that meant needing to be careful about handling translations of the (non-string) data, needing to carefully encode back to bytes before using. This was both an expensive pointless operation, and prone for error, so we went for strict bytes.

This was also one of the problems running the original pycbf tests - they didn't work in python 3 in either strict or non-strict mode, because they made different assumptions about whether things were bytes or str. https://github.com/dials/pycbf/commit/647ffcb5a80ea6148c4bc1d734ae6f7928a294ea was the dials-flavor pycbf rewrite of the tests to account for that. I don't know if your current problem running tests for pycbf/cbflib is related to this, but I think it depends on how you are building. I attempted the CMake build of CBFlib earlier today and it.. didn't work well; it still tried to use hdf5-1.8 and thought it was 0.9.6 (possibly because they forgot to update the version number: [yayahjb/cbflib@include/cbf.h](https://github.com/yayahjb/cbflib/blob/5232590ec6cc05bb2705acc3413ea598263854da/include/cbf.h#L262-L264)), and downloaded the 0.9.6 test data. The tests didn't pass on plain 0.9.6, so I don't know if that situation persists in 0.9.7.

**Reply by picca** [reply]
Hello, I investigated a bit more.

When you build dials with the bootstrap it build via cbflib_adaptbx a modified version of cbf , a cbflib_ext Python extension and pycbf from an embeded pycbf_wrap.c. I patched the build system to build  the cbdlib_ext with only the detectors extra part with a small amount of patch of the sources. the includes of cbf are not right for a system installation. I skipped the pycbf part and end up with the same problem observed previously, indeed the cbflib_adaptbx build a pycbf with the global `-DSWIG_PYTHON_STRICT_BYTE_CHAR` set which is not the case of the upstream cbflib.

I think that I will keep the official cbflib way of building the pycbf python module (less surprise for our users).

Now I need to modify dxtbx in order to have something working with this pycbf, so my question is do you think that this could solve the problem

https://github.com/cctbx/dxtbx/compare/1876d2316e30c29af3840dc8f6d743b3809115ee...ndevenish:string_cbf

thanks for your time

Fred

**Reply by picca** [reply]
I found an issue in the Debian build of cbflib... since we do not have nuweb, I used the m4/setup_py.m4 files to generate the setup.py for the Python3 pycbf file. but this file is different from the one in the pycbf.w

```m4
@o py3setup_py.m4 -i -t
@{
#
#  py3setup_py.m4
#

`# Import the things to build python binary extensions

from distutils.core import setup, Extension

# Make our extension module

e = Extension(''`_pycbf''`,
              sources = ["pycbf_wrap.c","../src/cbf_simple.c"],
         extra_compile_args=["-g", "-DSWIG_PYTHON_STRICT_BYTE_CHAR"],
         'm4_ifelse(regexlibdir,`NOREGEXLIBDIR',`library_dirs=["../solib/","../lib/"],',`library_dirs=["../solib/","../lib/","'regexlibdir`"],')`
         'm4_ifelse(regexlib,`',`libraries=["cbf"],', `m4_ifelse(regexlib2,`',`libraries=["cbf","'regexlib`"],',`libraries=["cbf","'regexlib`","'regexlib2`"],')'  )`
         include_dirs = ["../include","'hdf5_prefix`/include"] )
            
# Build it
setup(name="_pycbf",ext_modules=[e],)'
@}
```

so the new cbflib compile pycbf with the global STRICT flag... It is strange to add it for a few methods and then set it globally.

sorry for this :(((
